dashboard/src/components/AppView/AppControls/RollbackButton/RollbackButton.test.tsx

```diff
@@ -0,0 +1,158 @@
+import { shallow, ShallowWrapper } from "enzyme";
+import context from "jest-plugin-context";
+import * as React from "react";
+import * as ReactModal from "react-modal";
+import { IAppRepository, IChartVersion, IRelease } from "shared/types";
+import RollbackButton from ".";
+import SelectRepoForm from "../../../../components/SelectRepoForm";
+import RollbackDialog from "./RollbackDialog";
+
+const defaultProps = {
+  app: {
+    chart: {
+      metadata: {
+        name: "foo",
+        version: "1.0.0",
+      },
+    },
+  } as IRelease,
+  rollbackApp: jest.fn(),
+  getChartVersion: jest.fn(),
+  loading: false,
+  repos: [],
+  repo: {} as IAppRepository,
+  kubeappsNamespace: "kubeapps",
+  fetchRepositories: jest.fn(),
+  checkChart: jest.fn(),
+};
+
+it("stores the chart name and version in the state", () => {
+  const wrapper = shallow(<RollbackButton {...defaultProps} />);
+
+  expect(wrapper.state()).toMatchObject({ chartName: "foo", chartVersion: "1.0.0" });
+});
+
+function openModal(wrapper: ShallowWrapper) {
+  ReactModal.setAppElement(document.createElement("div"));
+  wrapper.setState({ modalIsOpen: true });
+  wrapper.update();
+}
+
+context("when opening the Modal", () => {
+  context("when there is no info about the chart", () => {
+    it("should render the SelectRepoForm", () => {
+      const wrapper = shallow(<RollbackButton {...defaultProps} chart={undefined} />);
+      openModal(wrapper);
+
+      expect(wrapper.find(SelectRepoForm)).toExist();
```

Comment:
> > In what situations will we not have info about the chart? You mention in the description that this can happen when the repository is not an AppRepository, but it's not clear whether this means that Kubeapps does not have this info (I mean, helm must have it, right?), or the dashboard just doesn't have it (or can't currently get it), or it's one of stable etc.?
> 
> That could happen in several occasions:
> 
>     * The user installs a chart using the `helm` CLI with a repository that doesn't exists in Kubeapps. Then we are not able to see where this application comes from.

In this situation, helm/tiller then knows the repo though doesn't it? EDIT: I think not - this is the source of my confusion. Helm just finds a chart with whatever repos it has access to, and doesn't (apparently) store the repo of the chart it finds for later reference?

> 
>     * The chart doesn't have an `appVersion`. This field is optional but we use it to verify if a chart belongs to a repository. What we do is checking if `chart: wordpress, version: 1.0.0, appVersion: 1.0.1` exists in the repo `stable`. If those three parameters match, we assume that chart comes from that repository. If we don't have that information we are not confident enough to make that assumption.

And this is all info which isn't necessary to roll back, but just to verify the current user is allowed to roll back? Seems odd (but I may not have my head around it). Why can't we assume that if the user has "write" access to the namespace in which the chart is deployed, then they can roll back and just let them do it? I think on upgrade it's a bit different because it requires access to the repo, but does a rollback require this at all? (not sure if the chart versions are stored by tiller, haven't checked).

[snip]

> 
> > Isn't the repo required just to get the auth info for the request? If so, what's the implication of allowing me, the user, to select which auth to use? (security thought: I could setup my own repo which captures any auth credentials, then rollback/upgrade a release, but use the dashboard to specify another repo with secrets I want to know... won't the request then be done by tiller with the secrets from the other repo, so I can capture them?)
> 
> The repo is required to retrieve the information about the chart. This is ultimately used to verify if the user `foo` can modify the deployment `bar` that is specified in the chart. This is not perfect though since we don't have a way to verify that the chart really belongs to that repo, we just verify that there is a chart with the same version/appversion is there.

As above, why isn't it enough that the user has sufficient access to the namespace, to enable roll back?

> 
> The user doesn't interact with any other auth info so I am not sure if that's the case you are describing.
> 
> > Also, this is related to #1110 I think? (in that, we're only needing to read repositories because we need the auth... if instead we sent the repository name and tiller used that to get the auth, we'd only ever need the repo name and would not need to fetch the repo/repos). We just need to always have a repo name - interested to know why we don't know the repo name.
> 
> We would still need to list the repos if we are not able to auto-retrieve the registry that contains the chart.

Only if we required the repo for authz though right? (Unlike an upgrade, a rollback doesn't require access to the repo per-se, since it's rolling back to a previous version of the chart which it has stored?)

Reply:
> Helm just finds a chart with whatever repos it has access to, and doesn't (apparently) store the repo of the chart it finds for later reference?

Right, that's the problem. The helm CLI downloads the chart locally and just send the manifests. Tiller (or the configmaps stored) has no clue of where those charts came from.

> Why can't we assume that if the user has "write" access to the namespace in which the chart is deployed, then they can roll back and just let them do it?

There is not a simple check to verify if a user has "write" access. Note that a chart can contain any resource, they can even escape from a namespace. If a chart has a ClusterRole or a CRD we need to verify that the user has permissions to modify those cluster-wide resources. Even if it's in a single namespace, a user can have permissions to modify configmaps but not secrets. We need to verify all the resources to avoid an escalation of privileges.

>  I think on upgrade it's a bit different because it requires access to the repo, but does a rollback require this at all? (not sure if the chart versions are stored by tiller, haven't checked).

The rollback operation doesn't need to access the registry but we need to do so to be able to get the resources the chart modifies. Ideally we could get this information from Tiller since those resources are specified in the configmap but for simplicity we are not doing that right now. That could be an improvement of this approach.